### PR TITLE
Markers to filter tests in new pipeline

### PR DESCRIPTION
**Pull request by jyejare**
This PR includes:

#### 1. Markers to filter following tests in robottelo:

- [x] API
- [x] CLI
- [x] UI
- [x] End To End
- [x] Longrun
- [x] RHAI

#### 2. Pytest Plugin that modifies the collection of these tests based on new following options to select the tests:

- [x] **--include-onprem-provisioning** => Selects on-premises provisioning tests 
- [x] **--include-libvirt** => Selects libvirt based content host tests 
- [x] **--include-external-auth** => Selects External Auth based tests 
- [x] **--include-vlan-networking** => Selects VLAN networking tests 

**Most IMP:** The tests of which inclusion is not selected using above option(s), then those tests will be deselected by default. That also means if the option(s) is not provided to pytest it will be considered as to deselect those tests by default.

### CODE REVIEW

**Comment by jyejare** [comment]
@mshriver Out of following new markers :
```
        "cloud_provisioning: Tests that runs on cloud Providers",
        "on_premises_provisioning: Tests that runs on on_premises Providers",
        "libvirt_discovery: Tests depends on Libvirt Provider for discovery",
        "libvirt_content_host: Tests depends on Libvirt Provider for content hosts",
        "auth_source: Tests depends on external auth sources for login to satellite"
```

Only `cloud_provisioning` marked provisioning tests can run in new pipeline, others had to be filtered out !

File: pytest_plugins/markers.py [hunk 1]
```diff
@@ -14,6 +14,11 @@ def pytest_configure(config):
         "upgrade: Upgrade tests",
         "run_in_one_thread: Sequential tests",
         "build_sanity: Fast, basic tests that confirm build is ready for full test suite",
+        "cloud_provisioning: Tests that runs on cloud Providers",
+        "on_premises_provisioning: Tests that runs on on_premises Providers",
+        "libvirt_discovery: Tests depends on Libvirt Provider for discovery",
+        "libvirt_content_host: Tests depends on Libvirt Provider for content hosts",
+        "auth_source: Tests depends on external auth sources for login to satellite"
```
Comment (mshriver) [inline pytest_plugins/markers.py:21]: I think this marker (and maybe the others eventually) should be applied to the fixtures for auth source entity creation.

This way the test author doesn't have to add the fixture, and apply marks when they've used the fixture.

Comment (jyejare) [reply pytest_plugins/markers.py:21]: @mshriver Cool ! Thanks will do !

Comment (jyejare) [reply pytest_plugins/markers.py]: Done !

**Comment by jyejare** [comment]
@mshriver Revisit again, I have pytest_collection hook now.

File: conftest.py [hunk 2]
```diff
@@ -3,6 +3,7 @@
 pytest_plugins = [
     # Plugins
     "pytest_plugins.rerun_rp.rerun_rp",
+    "pytest_plugins.new_pipeline",
```
Comment (jyejare) [inline conftest.py:6]: I can also rename this to satlab_infra or something else!

File: pytest_plugins/new_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+LOGGER = logging.getLogger('robottelo')
+
+
+def pytest_addoption(parser):
+    """Add options for pytest to collect tests than can run on SatLab infra"""
+    help_text = f'''
+        Filter out tests which cannot be run on a new pipeline
+
+        Usage: `pytest tests/foreman --new-pipeline`
```
Comment (jyejare) [inline pytest_plugins/new_pipeline.py:15]: Even this can be renamed to `--satlab-infra`.

**Comment by jyejare** [comment]
@omkarkhatavkar Can you please visit the PR and let me know what tests really need to filter out for new pipeline unless we set up some thing in SatLab ?

**Comment by omkarkhatavkar** [comment]
@jyejare you can add filters on the following tests as these are realm tests `test_single_sign_on_ldap_ipa_server, test_single_sign_on_ldap_ad_server, test_positive_create_with_idm_https`. RHSSO tests need a small modification to make sure that needs to be getting run on Satlab. That can be taken are in the robottelo soon.

**Comment by JacobCallahan** [comment]
@jyejare did you mean to include a pytest conversion in this as well?

File: pytest_plugins/markers.py (hunk 1 above)
Comment (jyejare) [reply pytest_plugins/markers.py]: @mshriver I noticed that markers assigned to fixtures are not playing any role is test selection/deselection atleast in collect only :(

Do you feel otherwise ?

**Comment by jyejare** [comment]
@mshriver The description of this PR along with the plugin changes to filter the tests based on `include` options is updated, please revisit.

**Comment by lpramuk** [comment]
if RHAI then VirtWho too ?

**Comment by jyejare** [comment]
@lpramuk The VirtWho has its own CI as well as infra and doesnt required to filtered out? :)

**Comment by jyejare** [comment]
@lpramuk @mshriver Before we merge this, we have to modify the old pipeline to include them all :) else they will be auto filtered out !

**Comment by mshriver** [comment]
@jyejare robottelo-ci is now running off of `6.9.freeze` branch, so it will not need a change to its arguments.

**Comment by jyejare** [comment]
@mshriver Ohh, worry free then.

File: tests/foreman/api/test_discoveredhost.py [hunk 4]
```diff
@@ -289,441 +256,481 @@ def discovered_host_cleanup():
...
-    :CaseImportance: Critical
-    """
+        :CaseImportance: High
+        """
+        name = gen_choice(list(valid_data_list().values()))
+        result = self._create_discovered_host(name)
 
+        entities.DiscoveredHost(id=result['id']).delete()
+        search = entities.DiscoveredHost().search(
+            query={'search': 'name == {}'.format(result['name'])}
```
Comment (omkarkhatavkar) [inline tests/foreman/api/test_discoveredhost.py:343]: F-String

File: tests/foreman/api/test_discoveredhost.py [hunk 5]
```diff
@@ -289,441 +256,481 @@ def discovered_host_cleanup():
...
+                discovered_host.auto_provision()
 
-    :CaseImportance: High
-    """
-    name = gen_choice(list(valid_data_list().values()))
-    result = _create_discovered_host(name)
+                # Assertions
+                provisioned_host = entities.Host(cfg).search(
+                    query={
+                        'search': 'name={}.{}'.format(
```
Comment (omkarkhatavkar) [inline tests/foreman/api/test_discoveredhost.py:483]: F -String

**Review by omkarkhatavkar — APPROVED** [review]
ACK on the selecting the external auth tests

File: pytest_plugins/markers.py (hunk 1 above)
Comment (mshriver) [reply pytest_plugins/markers.py]: @jyejare we would have to use a plugin like https://pypi.org/project/pytest-fixture-marker/ or pytest hooks to enforce this, out of the scope of your PR here.

File: pytest_fixtures/api_fixtures.py [hunk 6]
```diff
@@ -108,6 +108,7 @@ def module_domain(module_org, module_location):
     return entities.Domain(location=[module_location], organization=[module_org]).create()
 
 
+@pytest.mark.vlan_networking
```
Comment (mshriver) [inline pytest_fixtures/api_fixtures.py:111]: As you stated, marking the fixture directly doesn't carry to the tests, this has to be done with a pytest hook reading test item kwargs.

File: pytest_plugins/infra_dependent_markers.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# Custom infra dependent markers for selection and deselection of tests
+
+
+def pytest_configure(config):
+    """Register custom markers to avoid warnings."""
+    markers = []
+    markers.extend(infra_markers())
```
Comment (mshriver) [inline pytest_plugins/infra_dependent_markers.py:7]: Why all of this function calling to return a list?  You can just define a variable with the list at the module level, and iterate over it here without creating a new variable scope.

File: pytest_fixtures/satellite_auth.py [hunk 8]
```diff
@@ -227,6 +227,7 @@ def enroll_configure_rhsso_external_auth():
     run_command(cmd="systemctl restart httpd")
 
 
+@pytest.mark.external_auth
```
Comment (mshriver) [inline pytest_fixtures/satellite_auth.py:230]: Same here, these need to stay on the tests for now. We've got to use hooks to apply them through the fixtures.

File: pytest_plugins/marker_deselection.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    Collects and modifies tests collection based on pytest option to deselect tests for new infra
+    """
+    include_onprem_provision = config.getoption('include_onprem_provisioning', False)
+    include_libvirt = config.getoption('include_libvirt', False)
+    include_eauth = config.getoption('include_external_auth', False)
+    include_vlan = config.getoption('include_vlan_networking', False)
+    selected = []
+    deselected = []
+    # Cloud Provisioning Test can be run on new pipeline
+    for item in items:
```
Comment (mshriver) [inline pytest_plugins/marker_deselection.py:37]: I like the structure here, :+1:

File: tests/foreman/api/test_discoveredhost.py (hunk 4 above)
Comment (mshriver) [reply tests/foreman/api/test_discoveredhost.py:343]: I don't think the changes to this module for pytest conversion are meant to be part of this PR.

File: tests/foreman/longrun/test_provisioning_computeresource.py [hunk 10]
```diff
@@ -28,6 +28,8 @@
 from robottelo.utils.issue_handlers import is_open
 
 
+@pytest.mark.on_premises_provisioning
+@pytest.mark.vlan_networking
```
Comment (mshriver) [inline tests/foreman/longrun/test_provisioning_computeresource.py:32]: Same, these need to be moved to the tests for now.

File: tests/foreman/ui/test_computeresource.py [hunk 11]
```diff
@@ -37,6 +37,7 @@
     pytest.skip('skipping tests due to missing rhev settings', allow_module_level=True)
 
 
+@pytest.mark.on_premises_provisioning
```
Comment (mshriver) [inline tests/foreman/ui/test_computeresource.py:40]: Have to be moved to the tests for now.

File: tests/foreman/ui/test_remoteexecution.py [hunk 12]
```diff
@@ -71,6 +71,7 @@ def module_loc(module_org):
     return location
 
 
+@pytest.mark.libvirt_content_host
```
Comment (mshriver) [inline tests/foreman/ui/test_remoteexecution.py:74]: Have to move to tests.

**Review by mshriver — CHANGES_REQUESTED** [review]
My suggestion about marking fixtures was a bit misplaced, previous frameworks had specific hooks to handle that behavior, its not built into pytest.  Have to move a few of these around, but the hook and everything else looks solid.

File: pytest_fixtures/api_fixtures.py (hunk 6 above)
Comment (jyejare) [reply pytest_fixtures/api_fixtures.py:111]: @mshriver Does that mean, We had to create a hook or do we have any already in pytest.

File: pytest_plugins/marker_deselection.py (hunk 9 above)
Comment (jyejare) [reply pytest_plugins/marker_deselection.py:37]: :) Great !

File: pytest_fixtures/api_fixtures.py (hunk 6 above)
Comment (mshriver) [reply pytest_fixtures/api_fixtures.py:111]: We would have to create that hook in robottelo. There is a pytest plugin that applies marks automatically from fixtures, but it doesn't enforce the behavior you have in the hooks. Let's just directly mark tests for the time being.

File: tests/foreman/api/test_computeresource_libvirt.py [hunk 13]
```diff
@@ -34,6 +34,8 @@
 from robottelo.decorators import skip_if_not_set
 from robottelo.test import APITestCase
 
+pytestmark = pytest.mark.on_premises_provisioning
```
Comment (lpramuk) [inline tests/foreman/api/test_computeresource_libvirt.py:37]: Bad, not whole module requires provisioning set up, there are tier1,2 tests trying to insert 256 chars long strings - no vlans invloved

File: pytest_plugins/infra_dependent_markers.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    for marker in markers:
+        config.addinivalue_line("markers", marker)
+
+
+def infra_markers():
+    """ Register custom markers for infra dependent tests """
+    return [
+        "on_premises_provisioning: Tests that runs on on_premises Providers",
+        "libvirt_discovery: Tests depends on Libvirt Provider for discovery",
+        "libvirt_content_host: Tests depends on Libvirt Provider for content hosts",
```
Comment (lpramuk) [inline pytest_plugins/infra_dependent_markers.py:17]: In SatLab  there is only RHVM and OSP providers, I dont like names: 
`libvirt_*`

Comment (lpramuk) [reply pytest_plugins/infra_dependent_markers.py:17]: I'm OK with:
- `discovery`
- `content_host`

File: tests/foreman/cli/test_computeresource_libvirt.py [hunk 15]
```diff
@@ -49,6 +49,8 @@
 from robottelo.datafactory import parametrized
 from robottelo.decorators import skip_if_not_set
 
+pytestmark = pytest.mark.libvirt_content_host
```
Comment (lpramuk) [inline tests/foreman/cli/test_computeresource_libvirt.py:52]: Bad, not whole module requires provisioning set up, there are tier1,2 tests trying to insert 256 chars long strings - no vlans invloved

File: tests/foreman/api/test_computeresource_libvirt.py (hunk 13 above)
Comment (jyejare) [reply tests/foreman/api/test_computeresource_libvirt.py:37]: No, wrong, even though they are Tier1/2 , they are trying to connect to the libivrt CR from our old infra. 

I we run those test cases in new infra, they will fail to make contact and hence they have to be excluded with that marker.

File: tests/foreman/cli/test_computeresource_libvirt.py (hunk 15 above)
Comment (jyejare) [reply tests/foreman/cli/test_computeresource_libvirt.py:52]: No, wrong, even though they are Tier1/2 , they are trying to connect to the libivrt CR from our old infra. 

I we run those test cases in new infra, they will fail to make contact and hence they have to be excluded with that marker.

File: pytest_plugins/infra_dependent_markers.py (hunk 14 above)
Comment (jyejare) [reply pytest_plugins/infra_dependent_markers.py:9]: Right, The intuition of this PR is to filter the tests to exclude in SatLab :) and hence they are explicitly called with `libvirt_*`.

File: tests/foreman/api/test_discoveredhost.py (hunk 4 above)
Comment (jyejare) [reply tests/foreman/api/test_discoveredhost.py:343]: Right! That can taken care by Component owner whenever he/she will renew the tests.

File: pytest_fixtures/api_fixtures.py (hunk 6 above)
Comment (jyejare) [reply pytest_fixtures/api_fixtures.py:111]: Yeah! Marked all of the tests yday!

File: pytest_plugins/infra_dependent_markers.py (hunk 14 above)
Comment (jyejare) [reply pytest_plugins/infra_dependent_markers.py:9]: @lpramuk They are specially marked with `libvirt_` because discovery and content_host can also be the hosts of other compute resources like VMWare / RHEV etc, but these libvirt based tests are explicitly asking for old infra libvirt and hence we ll exclude only `libvirt_content_host` but not `rhev_content_host`.

I dont want put all of them in generic category of `content_host` so that excluding them would get difficult.

Comment (mshriver) [reply pytest_plugins/infra_dependent_markers.py:9]: @lpramuk correct, many of these libvirt tests will get removed and replaced by compute resource tests against compute resources that are officially supported (libvirt is not, but was used for testing convenience).

We want all of these libvirt tests marked and deselected by default so that they no longer run in the new pipelines.

File: conftest.py [hunk 16]
```diff
@@ -3,6 +3,7 @@
 pytest_plugins = [
     # Plugins
     'pytest_plugins.rerun_rp.rerun_rp',
+    "pytest_plugins.marker_deselection",
```
Comment (mshriver) [inline conftest.py:6]: infra_dependent_markers needs to be included here, the markers aren't getting registered with pytest.

Comment (mshriver) [reply conftest.py:6]: Also, please keep things consistent and use single quotes.

File: pytest_plugins/infra_dependent_markers.py [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+def pytest_configure(config):
+    """Register custom markers to avoid warnings."""
+    markers = [
+        "on_premises_provisioning: Tests that runs on on_premises Providers",
+        "libvirt_discovery: Tests depends on Libvirt Provider for discovery",
+        "libvirt_content_host: Tests depends on Libvirt Provider for content hosts",
+        "external_auth: External Authentication tests",
+        "vlan_networking: Tests depends on static predefined vlan networking etc",
+    ]
+    markers.extend(markers)
```
Comment (mshriver) [inline pytest_plugins/infra_dependent_markers.py:13]: Why is this line here, seems like it would double the list content?

Comment (mshriver) [reply pytest_plugins/infra_dependent_markers.py:13]: Yep, this does in fact double the registered markers: 

```
@pytest.mark.on_premises_provisioning: Tests that runs on on_premises Providers

@pytest.mark.libvirt_discovery: Tests depends on Libvirt Provider for discovery

@pytest.mark.libvirt_content_host: Tests depends on Libvirt Provider for content hosts

@pytest.mark.external_auth: External Authentication tests

@pytest.mark.vlan_networking: Tests depends on static predefined vlan networking etc

@pytest.mark.on_premises_provisioning: Tests that runs on on_premises Providers

@pytest.mark.libvirt_discovery: Tests depends on Libvirt Provider for discovery

@pytest.mark.libvirt_content_host: Tests depends on Libvirt Provider for content hosts

@pytest.mark.external_auth: External Authentication tests

@pytest.mark.vlan_networking: Tests depends on static predefined vlan networking etc

```

Comment (jyejare) [reply pytest_plugins/infra_dependent_markers.py:13]: Was in hurry .. and made mistakes ! :)

**Comment by jyejare** [comment]
@mshriver Resolved the quick but great comments ! :)

**Review by mshriver — APPROVED** [review]

**Comment by jyejare** [comment]
@mshriver Ready for merge ! Conflicts resolved !
